qmail-vdkim.8
.TH qmail-vdkim 8 "" openqmail
.SH NAME
qmail-vdkim \- verify DKIM signatures
.SH SYNOPSIS
.Bqmail-vdkim
.SH DESCRIPTION
\fBqmail-vdkim\fR have to plugin into the \fIqmail-smtpd\fR->\fIqmail-queue\fR
pipeline.
.P
\fBqmail-vdkim\fR checks the incoming message for a DKIM signature. On success it calls
an external library \fBlibqdkim\fR to verify the signature. It checks the result from
the external library and inserts a header '\fIX-Authentication-Result:\fR'.
.SH CONFIGURATION
.B qmail-vdkim 
uses several variables, which can be set in the environment too.
.TP 3
.IR DOVRFY
Switch verification ON or OFF (default: OFF). This have to be set \fIDOVRFY=1\fR in
\fI/usr/local/qmail/etc/qdkim.conf\fR explizit to enable verification.
.TP 3
.IR DKLIB
Name of the external library to use (default: \fBlibqdkim\fR).
.TP 3
.IR $DKQUEUE
This have to contain the name of the next program which should be called in the pipe.
Usually it is \fB/usr/local/qmail/bin/qmail-queue\fR (default).
.SH DEPENDENCIES
\fBqmail-vdkim\fR requires \fBunix2dos\fR from the package \fIdos2unix\fR.
.SH "AUTHOR"
Kai Peter (©2017), kp@openqmail.org
.SH "SEE ALSO"
qmail-sdkim(8), qdkim.conf(5), unix2dos(1)


CHANGELOG
commit f0246dee95471079078bc0cb9b1231ab7df20dab
Author: Kai Peter <kp@openqmail.org>
Date:   Sat May 13 00:02:48 2017 +0200

    removed dependancy 'unix2dos'

commit 258ab6af23bcb6506684566ab2d60dc7c8201dfd
Author: Kai Peter <kp@openqmail.org>
Date:   Fri May 12 09:57:32 2017 +0200

    dkimsign: corrected strings for c= tag

commit a57bafd4056759ca48a2f33aac170618a58cda03
Author: Kai Peter <kp@openqmail.org>
Date:   Thu May 11 19:20:14 2017 +0200

    cleanup dns.h

commit 946c7e207d39881d2bbc456783d9ade016e5975a
Author: Kai Peter <kp@openqmail.org>
Date:   Thu May 11 19:16:07 2017 +0200

    upped version to 20170511-beta1

commit 2997c025177f496e16448ce09bd727a1660aecd6
Merge: e7b3213 c3a5226
Author: Kai Peter <kp@openqmail.org>
Date:   Thu May 11 16:00:06 2017 +0200

    Merge pull request #2 from kp-org/dev
    
    merge of 'qdkim' into 'xdkim'

commit c3a5226ce8edcebfcc17481fb3be2a451f3fa061
Author: Kai Peter <kp@openqmail.org>
Date:   Thu May 11 15:55:16 2017 +0200

    clean-up, docs

commit ab48b27dcacb3ed370b556536915ff6b4220a07b
Author: Kai Peter <kp@openqmail.org>
Date:   Thu May 11 15:50:30 2017 +0200

    added libqdkim code

commit e47d1b44a28aa3f9bd009e23c832ab5613e25635
Author: Kai Peter <kp@openqmail.org>
Date:   Thu May 11 15:49:03 2017 +0200

    added config file

commit b49457386bfbfad3317968026e70697bff326fc6
Author: Kai Peter <kp@openqmail.org>
Date:   Wed May 10 10:19:25 2017 +0200

    qmail-sdkim: removed obsolete code, updated man page

commit 2f3a0d5a2f09982a1a606ef95bd5833298043f55
Author: Kai Peter <kp@openqmail.org>
Date:   Wed May 10 09:56:49 2017 +0200

    renamed 'mkdomainkey' to 'mkdkimkey', improved code and man page

commit 3fb114135c4dd95b235b97702f20998850052115
Author: Kai Peter <kp@openqmail.org>
Date:   Sun May 7 08:57:43 2017 +0200

    qmail-sdkim: more failsafe, usage of config file

commit 09e643ca2a0d961612e89f049a294135bcc0e81e
Author: Kai Peter <kp@openqmail.org>
Date:   Fri May 5 11:25:02 2017 +0200

    added Makefile and switched to 'make' build process

commit c579989a7e34e98c7cd5153bd7254eceaa6ba962
Author: Kai Peter <kp@openqmail.org>
Date:   Fri May 5 11:07:02 2017 +0200

    typo in mkdomainkey.8

commit e7b3213cac84b9a3be220680b2bf9878e98cd3a1
Merge: 939df0d 4cac84d
Author: Kai Peter <kp@openqmail.org>
Date:   Fri Feb 19 11:47:42 2016 +0100

    Merge pull request #1 from kp-org/dev
    
    first merge to master

commit 4cac84d5267b25b23d6021274e3f4273fd69a2ea
Author: Kai Peter <kp@openqmail.org>
Date:   Fri Feb 19 11:13:44 2016 +0100

    initial commit

commit 939df0d2726c2ea71c805d5cb150f1231aa3630e
Author: Kai Peter <kp@openqmail.org>
Date:   Thu Feb 18 12:59:10 2016 +0100

    Create README.md


########## BELOW IS THE CHANGELOG BEFORE USING GIT ##########

Changes: qmail-xdkim ****************************************

20170219: created git repo
20160218: qmail-vdkim-0.42-beta
          - added variable XDKIM
          - created man page qmail-vdkim.8
20160211: qmail-sdkim-0.44
          - added variable XDKIM
20160210: qmail-sdkim-0.43
          - added PARENTLEVEL for third and fourth level domains

Changes: libdkim/xdkim **************************************

20160426:
  - changed license to ISC
20160211:
  - removed unused DKIMSignGetSig function (commented out)
  - added man page
  - deleted WIN32 stuff
20160210: converted all files to UNIX format
  - added configure, Makefile.in
  - removed '#ifdef WIN32' (commented out)
  - changed default selector to "default"
  - unset "bardenhagen" identity and domain
20160208:
  - disabled requirement 'subject' (changed options to 1)
20160205:
  - strtok_r.patch
  - fix_warnings.patch
  - dos2unix Makefile
  - added '-lresolv' to LDFLAGS
  - applied corrected usage patch


qmail-vdkim.sh
#********************************************************************************
# DKIM verification for openqmail/eQmail/(net)qmail and derivatives             #
#                                                                               #
#  Author: Kai Peter (kp@openqmail.org), ©2014,©2017                            #
# Version: 0.55                                                                 #
# Licence: see file LICENSE                                                     #
#                                                                               #
# Description: Verification of DKIM signatures and writing of the header line   #
#              "Authentication-Results:".                                       #
#********************************************************************************
PATH="QMAILHOME/bin:$PATH"
# check for config file and include it
[ -f "QMAILHOME/etc/qdkim.conf" ] && . "QMAILHOME/etc/qdkim.conf"

# these files HAVE TO exists from *qmail, else - bummer
[ -f "$DKQUEUE" ] || DKQUEUE=`dirname $0`"/qmail-queue"
# do not use 'test -f' because $LOGGER could have params
[ "$LOGGER" ] || LOGGER="QMAILHOME/bin/splogger"

[ "$DOVRFY" ] || DOVRFY=0    # make sure it is initialized

DelTmpFiles() { rm -f $HEADER $inMsg $tmpMsg; }
DoLog() { echo "$LogMsg" | $LOGGER qmail-vdkim; }

# just for backwards compatibility (libqdkim is preferred)
[ "$DKLIB"  ] || DKLIB="QMAILHOME/bin/libqdkim"
if [ ! -f "$DKLIB" ] && [ "$DOVRFY" = "1" ] ; then
  LogMsg="warning: couldn't find $DKLIB - unable to verify message"
  DoLog ; DOVRFY=0 ; fi

if [ ! "$DOVRFY" = "1" ] ; then $DKQUEUE "$@"; exit 0 ; fi

#********************************************************************************
# try to verify DKIM signature(s) ***********************************************
# first set some default vars
[ -d "QMAILHOME/tmp" ] && TMPDIR="QMAILHOME/tmp" || TMPDIR="/tmp"
[ `uname -s | tr 'A-Z' 'a-z'` = "linux" ] && O="-f"    # ... and again BSD  >:(
HOSTNAME=`hostname "$O" | tr 'A-Z' 'a-z'`
DONEBY="$HOSTNAME"  # doesn't confirm with RfC6376 exatily
DOMAIN="unknown"    # initialize
VALID="fail"        # default assumption
inMsg=`mktemp -p "$TMPDIR" -t vdkim.XXXXXXXXXXXXXXXXXXX`
cat - >"$inMsg"     # save the message

# check for header 'DKIM-Signature'
cat $inMsg | grep "^DKIM-Signature" 2>&1 >/dev/null
if [ $? -ne 0 ] ; then
  LogMsg="qmail-vdkim: no DKIM signatures found." ; DoLog
  cat "$inMsg" | $DKQUEUE "$@" ; rm -f "$inMsg" ; exit 0 ; fi

tmpMsg=`mktemp -p "$TMPDIR" -t vdkim.XXXXXXXXXXXXXXXXXXX`
HEADER=`mktemp -p "$TMPDIR" -t vdkim.XXXXXXXXXXXXXXXXXXX`
# we need CRLF line endings for verification
QMAILHOME/bin/qdkimcrlf cat "$inMsg" >"$tmpMsg"

# redirection of stderr and pipe to /dev/null is a MUST!
$DKLIB -v $tmpMsg | grep 'Signature #[0-9]: Success' 2>&1 >/dev/null
if [ $? -eq 0 ] ; then VALID="pass (ok)" ; fi

# get the message body - just a coding reference here :))
#cat $tmpMsg | sed 's/\r//' | sed '1,/^$/d' > $BODY
# get the message headers
cat $tmpMsg | sed 's/\r//' | sed '/^$/q' | \
    sed -e :a -e '/^\n*$/{$d;N;ba' -e '}' > $HEADER
# get domain from "From:" header field
DOMAIN=$(cat $HEADER | grep ^[Ff]rom: | tr -d '\15' | cut -d@ -f2 | cut -d\> -f1)
# prepare new header field(s)
printf "X-Authentication-Results: $DOMAIN; dkim=$VALID; $DONEBY\n" | \
       fmt -w 78 -t > $HEADER
# concatenate new header field (DKIM-Verify) and message
cat "$HEADER" "$inMsg" | "$DKQUEUE" "$@"
EC=$?
LogMsg="qmail-vdkim: Verify DKIM for $DOMAIN: $VALID" ; DoLog
DelTmpFiles
exit $EC
